design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,138 @@
+**Release Target**: 0.25 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/8293
+
+**tl;dr:** _Expose Kafka header data in ksqlDB. Currently, headers are the only metadata in a Kafka record that is invisible to users._
+
+## Motivation and background
+
+So far, ksqlDB exposes the timestamp, partition and offset of Kafka records as pseudocolumns. The only metadata left that is invisible to ksqlDB users are [headers](https://kafka.apache.org/28/javadoc/org/apache/kafka/common/header/Headers.html). Kafka headers are a list of zero or more key-value pairs, where the keys are (not necessarily unique) strings and the values are byte arrays. Headers typically contain metadata about records, which can then be used for routing or processing (e.g. the header could store information about how to deserialize strings in the value). If ksqlDB exposes headers to the user, then these steps could be done in ksqlDB.
+
+## What is in scope
+* Syntax for accessing headers
+* Support for all query types except for `INSERT` queries
+* Including header columns in a data source's schema
+* Functions to help decode byte strings to primitive types (BYTES to STRING already exists, new functions would be conversions to INT, BIGINT and DOUBLE)
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values.
+```
+ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
+The type of a header backed column must be exactly `ARRAY<STRUCT<key STRING, value BYTES>>`. An error will be thrown if that is not the case.
+```
+ksql> CREATE STREAM B (my_headers ARRAY<BYTES> HEADERS);
+Error: Columns specified with the HEADERS keyword must be typed as ARRAY<STRUCT<key STRING, value BYTES>>.
+```
+Users can also mark a column with `HEADER(<key>)`, which will populate the column with the last header that matches the key. The data type must be `BYTES`
+```
+ksql> CREATE STREAM B (value BYTES HEADER(<key>));
+```
+If the key is not present in the header, then the value in the column will be `null`.
+
+Sources created with schema inferencing can have header columns, they would just have to be specified. For example, the statement `CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');` will create a stream with columns from Schema Registry, as well as one header column.
```

Comment:
> I don't think we should support schema inference this way

@spena to me the KLIP essentially suggests that schema inference is left alone if you define a column with the `HEADER(<key>)` or `HEADERS` keywords. Any column specified with that keyword is ignored from the perspective of schema registry: we're not trying to apply a SR schema to it or derive something in the output SR schema from it.

Reply:
I'm not sure I follow this section then. My understanding is that this would happen:

AVRO SR = (id INT, name STRING)
```
# with headers
ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');
ksql> DESCRIBE A;
(id INT, name STRING, my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS)

# without headers
ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>>) WITH (kafka_topic='a', value_format='avro');
ksql> DESCRIBE A;
(my_headers ARRAY<STRUCT<key STRING, value BYTES>>)
```

Is the above correct? If so, that's what I think shouldn't happen. I think it will make the user confused about it. I'm not sure how it should be, but I'm just concerned that will generate other problems. I think too that we could solve it in the same way we will solve (not sure how) other column attributes in the column definition. For example:

AVRO SR = (id INT, name STRING)
```
CREATE STREAM A (id INT, name STRING NOT NULL) WITH (kafka_topic='a', value_format='avro');
```

 The above is just an example that if we have to include attributes that are not in SR, then we have to define the schema manually. Or have another alternative for SR inference with extra column attributes in specific columns.

Reply:
@spena That is correct. The behaviour would be similar to [partial injections](https://docs.ksqldb.io/en/latest/operate-and-deploy/schema-registry-integration/#with-partial-schema-inference).

The difference between HEADERS and other attributes like NOT NULL is that the data in headers are not included in SR, while a column with the NOT NULL attribute can still be included in a schema from SR.

Reply:
I agree that partial projection can be confusing to users, but I don't see that adding headers make the situation worse? @spena can you elaborate on your concern? Why does adding headers make it more confusing for users from your POV?

Also not sure how, `NOT NULL` et al. relate? I don't see any connection? Can you elaborate?

Reply:
I see. I wasn't aware of the partial injections feature. In that case, then the UX is the same with HEADERS than with having KEY as the only column and infer the value columns. 

@mjsax The relation with `NOT NULL` it was just an example of adding attributes next to the column. In this case we have 3 type of attributes that specify a different scope of the column (KEY, HEADERS and VALUE). I was just getting confused.